descs,catkv: rewrite Collection backend

postamar opened the pull request:
This rewrite was motivated by the upcoming need to validate descriptors
up to different levels depending on why they were being read. Doing so
requires tracking the validation level in a more granular fashion within
the Collection's caches. Trying to make this work within the existing
scheme where the uncommitted and stored descriptors are all in the same
layer turned out to be painful and hard to reason about.

This commit does several things:
1. The uncommitted descriptors are spun off into their own distinct layer
again.
2. The stored descriptors now mirror what's in storage and so that whole
layer could effectively be moved to catkv.
3. Execution flow through the layers in the Collection is now
straightforwardly implemented and easy to reason about. Validation and
hydration now take place in unique, well-defined steps at the very end
of a lookup, be it by name or by ID.
3. Instances of this new stored descriptors layer are now also used by
the lease manager, the Direct() interface provided by the Collection,
etc.
4. As a result the logic in catkv could be significantly cleaned up.
In particular, lookups for descriptor validation are significantly
less convoluted now.
5. The SystemNamespaceCache object was also moved to catkv and now
supports caching descriptors, and so was renamed to
SystemCatalogCache.

Despite significant rewriting of the inner workings of the descriptors
collection, this commit should not functionally change anything.

Informs #85263.

Release justification: low-risk, high-benefit change.
Release note: None

ajwerner committed:
sql/catalog/descs: combine synthetic descriptors with stored descriptors

This allows synthetic descriptors to show up in virtual tables.

Release justification: part of bug fix.

Release note: None

ajwerner committed:
sql: rework some internal executor logic related to synthetic descripotrs

The pre-existing logic would reset the synthetic descriptors whenever anything
was run on behalf of the internal planner. That's not what we want. We
prevent that by attaching the logic to reset the synthetic descriptors to
whether the internal executor has attached synthetic descriptors, which
it never has when using the declarative schema changer.

Release justification: part of bug fix

Release note: None

Commit:
descs,catkv: rewrite Collection backend

This rewrite was motivated by the upcoming need to validate descriptors
up to different levels depending on why they were being read. Doing so
requires tracking the validation level in a more granular fashion within
the Collection's caches. Trying to make this work within the existing
scheme where the uncommitted and stored descriptors are all in the same
layer turned out to be painful and hard to reason about.

This commit does several things:
1. The uncommitted descriptors are spun off into their own distinct layer
   again.
2. The stored descriptors now mirror what's in storage and so that whole
   layer could effectively be moved to catkv.
3. Execution flow through the layers in the Collection is now
   straightforwardly implemented and easy to reason about. Validation and
   hydration now take place in unique, well-defined steps at the very end
   of a lookup, be it by name or by ID.
3. Instances of this new stored descriptors layer are now also used by
   the lease manager, the Direct() interface provided by the Collection,
   etc.
4. As a result the logic in catkv could be significantly cleaned up.
   In particular, lookups for descriptor validation are significantly
   less convoluted now.
5. The SystemNamespaceCache object was also moved to catkv, renamed to
   SystemDatabaseCache and now caches based on the cluster version.

Despite significant rewriting of the inner workings of the descriptors
collection, this commit should not functionally change anything.

Informs #85263.

Release justification: low-risk, high-benefit change.
Release note: None

Commit:
backupinfo: fix bug in writing names to metadata

Recent changes to the descriptors collection have uncovered some
brittleness in the logic for writing namespace entries for dropped
descriptors when creating backups.

This commit makes it:
- write deletion entries for descriptors other than tables, preventing
  possible catalog corruption.
- throw a better error message when attempting to write two identical
  keys consecutively.

Release justification: low-risk change
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87067)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
This is nice. The only major hesitation I have is around the `SystemDatabaseCache`. The rest of my comments are minor. 

> Reviewed 1 of 1 files at r1, 34 of 34 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-:-NAklXfY7QI1Zj5w_iXK:b-mg68qm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar and @rhu713)

___
*[`pkg/sql/catalog/descs/descriptor.go` line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkkEOzBJOrWKfXRfN9:-NAkkEOzBJOrWKfXRfNA:bxxbbxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/descriptor.go#L225)):*
> ```Go
> 	//
> 	// TODO(ajwerner): More generally leverage this set of kv descriptors on
> 	// the resolution path.
> ```

I think I'd say Jason achieved this TODO. 



_Code quote:_
```Go
	// TODO(ajwerner): More generally leverage this set of kv descriptors on
	// the resolution path.
```
___
*[`pkg/sql/catalog/descs/descriptor.go` line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkdg5B1imosj3i-SCe:-NAkdg5B1imosj3i-SCf:bxw47h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/descriptor.go#L338)):*
> ```Go
> // and (2) if mutable descriptors are requested, these are present in the
> // uncommitted descriptors layer.
> // Known validation levels can be provided via validationLevels, otherwise when
> ```

nit: my brain had a hard time parsing `otherwise when nil safe defaults are used instead`. Can you sprinkle in a comma or two?



_Code quote:_
```Go
// Known validation levels can be provided via validationLevels, otherwise when
// nil safe defaults are used instead. In any case, after validation is
```
___
*[`pkg/sql/catalog/descs/descriptor.go` line 348 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkgonMAowVaK5nJVhj:-NAkgonMAowVaK5nJVhk:b7nw1sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/descriptor.go#L348)):*
> ```Go
> 	validationLevels []catalog.ValidationLevel,
> ) error {
> 	if validationLevels == nil {
> ```

add a defensive check that if `validationLevels` is not `nil` that it has the same length as `descs`?
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkhK0d2bfC9J2Jqh6q:-NAkhK0d2bfC9J2Jqh6r:bcza9h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/uncommitted_descriptors.go#L181)):*
> ```Go
> 		return errors.Wrap(err, "Memory usage exceeds limit for uncommittedDescriptors")
> 	}
> 	ud.mutable.Upsert(mut, true /* skipNameMap */)
> ```

why are we unconditionally skipping the name map? It feels like if we're going to always skip the nameMap we should use a data structure that does not have one. 
___
*[`pkg/sql/catalog/internal/catkv/stored_catalog.go` line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkkjNT-4tDdfai5DYB:-NAkkjNT-4tDdfai5DYC:b-kcwqd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/stored_catalog.go#L38)):*
> ```Go
> 
> 	// cache mirrors the descriptors in storage.
> 	// This map does not store descriptors by name.
> ```

I said it elsewhere: let's export an object from `nstree` which wraps the existing `byIDMap` and avoid the confusion this thing is causing
___
*[`pkg/sql/catalog/internal/catkv/system_database_cache.go` line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAki5qtD-OMQIRAjYHh:-NAki5qtD-OMQIRAjYHi:be6ri9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/system_database_cache.go#L34)):*
> ```Go
> // the lease manager to cache all system table IDs.
> type SystemDatabaseCache struct {
> 	syncutil.RWMutex
> ```

I don't think it's right to export the mutex methods. The style here is generally to embed the mutex inside of an anonymous, embedded struct. I think that would work nicely here.

```
type SystemDatabaseCache struct {
	mu struct {
		syncutil.RWMutex
		m map[roachpb.Version]*nstree.MutableCatalog
        }
}
```
___
*[`pkg/sql/catalog/internal/catkv/system_database_cache.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkizM-55G0Tri6PESG:-NAkizM-55G0Tri6PESH:b-4r7p9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/system_database_cache.go#L35)):*
> ```Go
> type SystemDatabaseCache struct {
> 	syncutil.RWMutex
> 	m map[roachpb.Version]*nstree.MutableCatalog
> ```

The semantics of the versions in use here are interesting. I think that right now, it doesn't really matter because we never rename something in the system database and we only use this for caching names, right? 
___
*[`pkg/sql/catalog/internal/catkv/system_database_cache.go` line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkiJ_4AW_z_HsuZ-wF:-NAkiJ_4AW_z_HsuZ-wG:b-hrzsky) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/system_database_cache.go#L107)):*
> ```Go
> // Anything not pertaining to the system database is ignored.
> // Presently, system descriptors are all ignored and only namespace entries are
> // used.
> ```

what's your vision here? My plan for 23.1 is to start treating system descriptors, to the extent possible, like regular descriptors and not to make them more special. We'll need this to allow for deeper configuration of multi-region primitives



_Code quote:_
```Go
// Presently, system descriptors are all ignored and only namespace entries are
// used.
```


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
Thanks for the review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-:-NAo6LBl3Y-V_TXG3dJQ:b-9qdoyq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @rhu713)

___
*[`pkg/sql/catalog/descs/descriptor.go` line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkkEOzBJOrWKfXRfN9:-NAo5r4aAL1VPF9DBOiV:b-v6p8xk) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/descriptor.go#L225)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think I'd say Jason achieved this TODO. 
</blockquote></details>

Removing.
___
*[`pkg/sql/catalog/descs/descriptor.go` line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkdg5B1imosj3i-SCe:-NAo5siY3kL1M_mafAzh:b-17wys6) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/descriptor.go#L338)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: my brain had a hard time parsing `otherwise when nil safe defaults are used instead`. Can you sprinkle in a comma or two?
</blockquote></details>

Rewriting.
___
*[`pkg/sql/catalog/descs/descriptor.go` line 348 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkgonMAowVaK5nJVhj:-NAo5u3uBdEMKSe3ohzp:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/descriptor.go#L348)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

add a defensive check that if `validationLevels` is not `nil` that it has the same length as `descs`?
</blockquote></details>

Will do.
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkhK0d2bfC9J2Jqh6q:-NAo5vYf7I5UtMMXuldl:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/descs/uncommitted_descriptors.go#L181)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why are we unconditionally skipping the name map? It feels like if we're going to always skip the nameMap we should use a data structure that does not have one. 
</blockquote></details>

Will do.
___
*[`pkg/sql/catalog/internal/catkv/stored_catalog.go` line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkkjNT-4tDdfai5DYB:-NAo5xhkBVP8jREFyco-:b-hwswa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/stored_catalog.go#L38)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I said it elsewhere: let's export an object from `nstree` which wraps the existing `byIDMap` and avoid the confusion this thing is causing
</blockquote></details>

Will do. Agreed.
___
*[`pkg/sql/catalog/internal/catkv/system_database_cache.go` line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAki5qtD-OMQIRAjYHh:-NAo6-XmBhYJMN5mZz5D:b-wzck3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/system_database_cache.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think it's right to export the mutex methods. The style here is generally to embed the mutex inside of an anonymous, embedded struct. I think that would work nicely here.

```
type SystemDatabaseCache struct {
	mu struct {
		syncutil.RWMutex
		m map[roachpb.Version]*nstree.MutableCatalog
        }
}
```

</blockquote></details>

Whoops, right. Will do.
___
*[`pkg/sql/catalog/internal/catkv/system_database_cache.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkizM-55G0Tri6PESG:-NAo61xL0uWRFZq3lrtS:b-1nqt99) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/system_database_cache.go#L35)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The semantics of the versions in use here are interesting. I think that right now, it doesn't really matter because we never rename something in the system database and we only use this for caching names, right? 
</blockquote></details>

Indeed it's just me being defensive. Adding commentary.
___
*[`pkg/sql/catalog/internal/catkv/system_database_cache.go` line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NAkiJ_4AW_z_HsuZ-wF:-NAo65iqEVMKEFFOyX73:bnyt9q5) ([raw file](https://github.com/cockroachdb/cockroach/blob/baefec5c323b22263fd37e536ffa00bf186a5efc/pkg/sql/catalog/internal/catkv/system_database_cache.go#L107)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what's your vision here? My plan for 23.1 is to start treating system descriptors, to the extent possible, like regular descriptors and not to make them more special. We'll need this to allow for deeper configuration of multi-region primitives
</blockquote></details>

My vision is the same as yours. Adding commentary to explain why we don't actually cache system descriptors other than the system DB.


<!-- Sent from Reviewable.io -->

Commit:
nstree: rename Map to NameMap, add and adopt IDMap

This commit adds a new nstree.Map-like data structure without any
by-name operations.

Release justification: low-risk change which does not change
functionality
Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> my comments are nits, but in general, I'd prefer we not export methods unless we need them to conform to some interface explicitly. I spent a little while looking around for such an interface. 

The internal executor change I sort of get, but I don't fully get. I feel like we can tighten up the code and commentary around that change. 

> Reviewed 3 of 34 files at r3, 19 of 34 files at r4, 8 of 17 files at r5, 1 of 2 files at r6, 9 of 11 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-:-NApXZcS0lS2KaiMypxO:b-ldv6iy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @rhu713)

___
*[`-- commits` line 69 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NApYq4T91n3nAMgDau7):*
> which it never has when using the planner's state.

Is that true? Consider:

https://github.com/cockroachdb/cockroach/blob/c6f2617b48c1c00927c677cff4678e75c786f681/pkg/sql/backfill.go#L2641
https://github.com/cockroachdb/cockroach/blob/c6f2617b48c1c00927c677cff4678e75c786f681/pkg/sql/backfill.go#L2436
https://github.com/cockroachdb/cockroach/blob/17cfa7c7b7cb37d128a5f339d1e414a8ba83e3cc/pkg/sql/table.go#L298



_Code quote:_
```
  prevent that by attaching the logic to reset the synthetic descriptors to
  whether the internal executor has attached synthetic descriptors, which
  it never has when using the planner's state.
```
___
*[`-- commits` line 5 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NApXso18ia3b4iYrhir):*
I anticipate this all worked because the backups don't store dropped descriptors. Does this commit fix any known bug? Should it be in this PR or should it be pulled out into a separate one such that it might be tested?



_Code quote:_
```
  Recent changes to the descriptors collection have uncovered some
  brittleness in the logic for writing namespace entries for dropped
  descriptors when creating backups.
```
___
*[`pkg/sql/internal.go` line 337 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NApZLTlEuqO1BaFusry:-NApZLTlEuqO1BaFusrz:baoytam) ([raw file](https://github.com/cockroachdb/cockroach/blob/567c7fcf5651e4945e2a228232df4b9218a53dce/pkg/sql/internal.go#L337)):*
> ```Go
> 			ex.extraTxnState.descCollection = ie.extraTxnState.descCollection
> 			ex.extraTxnState.fromOuterTxn = true
> 			ex.extraTxnState.shouldResetSyntheticDescriptors = len(ie.syntheticDescriptors) > 0
> ```

can you add more commentary here?
___
*[`pkg/sql/internal.go` line 334 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NApRu4c9IItqkMhgXWT:-NApRu4d58pir8sbLPFM:b4gexoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bbc73bc6697414f406b15fd24cf8982471fe6/pkg/sql/internal.go#L334)):*
> ```Go
> 	// and job collections from the caller.
> 	postSetupFn := func(ex *connExecutor) {
> 		if ie.extraTxnState != nil && ie.extraTxnState.descCollection != nil {
> ```

what's the case where `extraTxnState` is not nil but `descCollection` is? Also, while you're here, if you have to do another revision, rename to `descsCollection`?
___
*[`pkg/sql/catalog/descs/synthetic_descriptors.go` line 23 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NApXEIbEHLZ9j7gdwJU:-NApXEIbEHLZ9j7gdwJV:b-92m0wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bbc73bc6697414f406b15fd24cf8982471fe6/pkg/sql/catalog/descs/synthetic_descriptors.go#L23)):*
> ```Go
> }
> 
> func (sd *syntheticDescriptors) Add(desc catalog.Descriptor) {
> ```

why are all these methods now exported?
___
*[`pkg/sql/catalog/descs/uncommitted_descriptors.go` line 63 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/87067#-NApVG-oEYsp3_ESjw92:-NApVG-oEYsp3_ESjw93:b8i5n6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bbc73bc6697414f406b15fd24cf8982471fe6/pkg/sql/catalog/descs/uncommitted_descriptors.go#L63)):*
> ```Go
> 
> // Reset zeroes the object for re-use in a new transaction.
> func (ud *uncommittedDescriptors) Reset(ctx context.Context) {
> ```

Why are all of the methods on this unexported struct exported? It doesn't seem like it's embedded anywhere


<!-- Sent from Reviewable.io -->

Commit:
descs: method name and signature changes

This commit doesn't otherwise change anything.

Release justification: very low-risk change
Release note: None
